installers/msi-language/StateDeploy/CustomAction.cs

```diff
@@ -122,21 +123,62 @@ private static ActionResult _installStateTool(Session session, out string stateT
             string zipURL = stateURL + info.version + "/" + paths.ZipFile;
             session.Log(string.Format("Downloading zip file from URL: {0}", zipURL));
             Status.ProgressBar.StatusMessage(session, "Downloading State Tool...");
-            try
+
+            var tokenSource = new CancellationTokenSource();
+            var token = tokenSource.Token;
+
+            Task incrementTask = Task.Run(() =>
             {
-                RetryHelper.RetryOnException(session, 3, TimeSpan.FromSeconds(2), () =>
+                if (token.IsCancellationRequested)
                 {
-                    var client = new WebClient();
-                    client.DownloadFile(zipURL, zipPath);
-                });
+                    token.ThrowIfCancellationRequested();
+                }
+                for (int i = 0; i <= 50; i++)
+                {
+                    if (token.IsCancellationRequested)
+                    {
+                        token.ThrowIfCancellationRequested();
+                    }
+                    Status.ProgressBar.Increment(session, 1);
+                    Thread.Sleep(150);
+                }
+            });
+
+            Task<ActionResult> downloadTask = Task.Run(() =>
+            {
+                try
+                {
+                    RetryHelper.RetryOnException(session, 3, TimeSpan.FromSeconds(2), () =>
+                    {
+                        var client = new WebClient();
+                        client.DownloadFile(zipURL, zipPath);
+                    });
+                }
+                catch (WebException e)
+                {
+                    string msg = string.Format("Encountered exception downloading state tool zip file. URL to zip file: {0}, path to save zip file to: {1}, exception: {2}", zipURL, zipPath, e.ToString());
+                    session.Log(msg);
+                    new NetworkError().SetDetails(session, e.Message);
+                    return ActionResult.Failure;
+                }
+
+                return ActionResult.Success;
+            });
+
+            try
+            {
+                incrementTask.Wait();
+            } catch (OperationCanceledException)
+            {
+                session.Log("Increment progress bar was cancelled");
             }
-            catch (WebException e)
+
+            ActionResult result = downloadTask.Result;
```

Comment:
I think here we need to add a `downloadTask.Wait()` first, in case the download task took longer than the `incrementTask`.

Reply:
See comment above